presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergHiveMetadata.java

```diff
@@ -161,15 +162,28 @@ private Optional<SystemTable> getRawSystemTable(ConnectorSession session, Schema
     }
 
     @Override
-    public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> schemaName)
+    public List<SchemaTableName> listTables(ConnectorSession session, String schemaName)
     {
         MetastoreContext metastoreContext = new MetastoreContext(session.getIdentity(), session.getQueryId(), session.getClientInfo(), session.getSource(), Optional.empty(), false, HiveColumnConverterProvider.DEFAULT_COLUMN_CONVERTER_PROVIDER);
-        return metastore
-                .getAllTables(metastoreContext, schemaName.get())
-                .orElseGet(() -> metastore.getAllDatabases(metastoreContext))
-                .stream()
-                .map(table -> new SchemaTableName(schemaName.get(), table))
-                .collect(toList());
+        ImmutableList.Builder<SchemaTableName> tableNames = ImmutableList.builder();
+        for (String schema : listSchemas(session, schemaName)) {
+            for (String tableName : metastore
+                    .getAllTables(metastoreContext, schemaName)
+                    .orElseGet(() -> metastore.getAllDatabases(metastoreContext))
+                    .stream()
+                    .collect(toList())) {
+                tableNames.add(new SchemaTableName(schema, tableName));
+            }
+        }
+        return tableNames.build();
+    }
+
+    private List<String> listSchemas(ConnectorSession session, String schemaNameOrNull)
+    {
+        if (schemaNameOrNull == null) {
+            return listSchemaNames(session);
+        }
+        return ImmutableList.of(schemaNameOrNull);
```

Comment:
Shouldn't this be empty rather than a list of null?

Reply:
If "schemaNameOrNull" is null, we will be returning a list of all schema names. And if it is not null, we are returning a list with just this value - which is non-null.

Reply:
IMO it would make more sense to structure it like this:

`schemaName` is optional.  If it's not present, then we should list schemas, and annotate table names for each schema.  If it's present, we don't list schemas, we simply list tables within the schema.

The way it's written is hard to grok.  We're saying, firstly, if schema is null, take that as an indication that we should list schema names.  Let's make it explicit: if it's optional and not present, this means we need to list schemas.  If it's present, don't call `listSchemas`--there is no listing because there's already a value present.

Does this make sense?

Reply:
Something like this, written using streams:

```java
    @Override
    public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> schemaName)
    {
        MetastoreContext metastoreContext = new MetastoreContext(session.getIdentity(), session.getQueryId(), session.getClientInfo(), session.getSource(), Optional.empty(), false, HiveColumnConverterProvider.DEFAULT_COLUMN_CONVERTER_PROVIDER);
        List<String> schemaNames = schemaName
                .map(ImmutableList::of)
                .orElseGet(() -> ImmutableList.copyOf(listSchemaNames(session)));
        return schemaNames.stream()
                .flatMap(schema -> metastore
                        .getAllTables(metastoreContext, schema)
                        .orElseGet(() -> metastore.getAllDatabases(metastoreContext))
                        .stream()
                        .map(table -> new SchemaTableName(schema, table)))
                .collect(toImmutableList());
    }
    ```